### PR TITLE
Postgres querying etc.

### PR DESCRIPTION
**Pull request by enola-dkfz**

### CODE REVIEW

**Comment by enola-dkfz** [comment]
there's a security issue in sqlx, its version needs to be increased to at least 0.8.1, that's the last version ATM, they have messed up and it doesn't compile
additionally, sqlx-pgrow-serde depends on that messed up version of sqlx

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,10 @@
+# Samply.Focus v0.7.0 2024-09-24
+
+## Major changes
+* PostgreSQL support added
```
Comment (TKussel) [inline CHANGELOG.md:4]:
```suggestion
# Samply.Focus v0.7.0 2024-09-24

In this release, we are extending the supported data backends beyond CQL-enabled FHIR stores. We now support PostgreSQL as well. Usage instructions are included in the Readme.

## Major changes
* PostgreSQL support added
```

Please adapt date before merging and add a couple of lines rgarding usage/feature flags in the readme.

File: Cargo.toml [hunk 2]
```diff
@@ -1,41 +1,47 @@
...
 # Global variables
 once_cell = "1.18"
 
 # Command Line Interface
-clap = { version = "4", default_features = false, features = ["std", "env", "derive", "help", "color"] }
+clap = { version = "4", default-features = false, features = ["std", "env", "derive", "help", "color"] }
+
+# Query via SQL
+sqlx = { version = "0.8.2", features = [ "runtime-tokio", "postgres", "macros", "chrono", "rust_decimal", "uuid"], optional = true } 
+kurtbuilds_sqlx_serde = { version = "0.3.2", features = [ "json", "decimal", "chrono", "uuid"], optional = true }
```
Comment (TKussel) [inline Cargo.toml:37]: You wanted to replace that, right?

File: README.md
Comment (TKussel) [inline README.md:1]: Please add a "How to" for Postgres, as it needs to be compiled with a different feature flag

File: src/errors.rs [hunk 3]
```diff
@@ -55,11 +56,20 @@ pub enum FocusError {
...
     MissingExporterTaskType,
+    #[error("Cannot connect to database: {0}")]
+    CannotConnectToDatabase(String),
+    #[error("QueryResultBad: {0}")]
+    QueryResultBad(String),
+    #[error("Query not allowed: {0}")]
+    QueryNotAllowed(String),
+    #[cfg(feature = "query-sql")]
+    #[error("Error executing query: {0}")]
+    ErrorExecutingQuery(sqlx::Error),
```
Comment (TKussel) [inline src/errors.rs:72]: That error can be useful for blaze as well, no? Why is it hidden behind the feature gate?

File: src/db.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+include!(concat!(env!("OUT_DIR"), "/sql_replace_map.rs"));
+
+pub async fn get_pg_connection_pool(pg_url: &str, max_attempts: u32) -> Result<PgPool, FocusError> {
+    info!("Trying to establish a PostgreSQL connection pool");
+
+    tryhard::retry_fn(|| async {
+        info!("Attempting to connect to PostgreSQL");
+        PgPoolOptions::new()
+            .max_connections(10)
```
Comment (TKussel) [inline src/db.rs:23]: In my opinion, that should not be hard coded.

File: src/db.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+pub struct SqlQuery {
+    pub payload: String,
+}
+
+include!(concat!(env!("OUT_DIR"), "/sql_replace_map.rs"));
+
+pub async fn get_pg_connection_pool(pg_url: &str, max_attempts: u32) -> Result<PgPool, FocusError> {
+    info!("Trying to establish a PostgreSQL connection pool");
+
+    tryhard::retry_fn(|| async {
```
Comment (TKussel) [inline src/db.rs:20]: Thanks for implementing exponential backoff!

File: src/db.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            .max_connections(10)
+            .connect(pg_url)
+            .await
+            .map_err(|e| {
+                warn!("Failed to connect to PostgreSQL: {}", e);
+                FocusError::CannotConnectToDatabase(e.to_string())
+            })
+    })
+    .retries(max_attempts)
+    .exponential_backoff(Duration::from_secs(2))
```
Comment (TKussel) [inline src/db.rs:32]: Is that bounded, i.e., exponentially increasing the backoff time up to an upper bound? If not, consider doing that (up to 30 seconds or so).

File: src/db.rs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        .map_err(FocusError::ErrorExecutingQuery)
+}
+
+pub async fn process_sql_task(pool: &PgPool, key: &str) -> Result<Vec<PgRow>, FocusError> {
+    debug!("Executing query with key = {}", &key);
+    let sql_query = SQL_REPLACE_MAP.get(&key);
+    let Some(query) = sql_query else {
+        return Err(FocusError::QueryNotAllowed(key.into()));
+    };
+    debug!("Executing query {}", &query);
```
Comment (TKussel) [inline src/db.rs:49]: Do the query or the key potentially include sensitive information? If yes, please do not log.

**Review by TKussel — CHANGES_REQUESTED** [review]

File: src/db.rs (hunk 7 above)
Comment (enola-dkfz) [reply src/db.rs:49]: This is a question for @davidmscholz

File: src/db.rs (hunk 6 above)
Comment (enola-dkfz) [reply src/db.rs:32]: Is max_attempts not enough?

File: src/errors.rs (hunk 3 above)
Comment (Threated) [reply src/errors.rs:72]: This Error contains an `sqlx::Error` so it can't be used otherwise. It should probably be renamed.

**Review by Threated — APPROVED** [review]

File: Cargo.toml (hunk 2 above)
Comment (enola-dkfz) [reply Cargo.toml:37]: I didn't need to in the end, @davidmscholz is OK with floats becoming strings

File: CHANGELOG.md (hunk 1 above)
Comment (enola-dkfz) [reply CHANGELOG.md:4]: But we have had EUCAIM querying as well

File: README.md
Comment (enola-dkfz) [reply README.md:1]: here: https://github.com/samply/focus/pull/171

File: src/errors.rs (hunk 3 above)
Comment (enola-dkfz) [reply src/errors.rs]: Renamed

File: src/db.rs (hunk 7 above)
Comment (enola-dkfz) [reply src/db.rs:49]: @davidmscholz said in the review that the query shouldn't contain anything sensitive

File: src/db.rs (hunk 5 above)
Comment (enola-dkfz) [reply src/db.rs:20]: ❤️

**Review by TKussel — APPROVED** [review]

File: src/db.rs (hunk 4 above)
Comment (TKussel) [reply src/db.rs:23]: While I am not a fan of the growing forrest of command line parameters, I think that would be the most "proper" way. Or pull it out into a constant for the time being.

File: src/db.rs (hunk 6 above)
Comment (TKussel) [reply src/db.rs:32]: Usually, exponential backoff uses an exponent of 2, so the wait times can grow unreasonably large even with "moderate" retries. Here with 10 retries and 2 seconds start delay, the last attempt would wait for 17 minutes. I think this is ok, just a potential footgun while tweaking the parameters.

**Review by davidmscholz — APPROVED** [review]
Querying with SQL worked as expected for the (basic) datatypes I tested.

File: src/db.rs (hunk 4 above)
Comment (enola-dkfz) [reply src/db.rs:23]: don't worry, it's behind a feature

File: CHANGELOG.md (hunk 1 above)
Comment (enola-dkfz) [reply CHANGELOG.md]: I just meant the formulation, but I did it as you suggested.

File: src/db.rs (hunk 4 above)
Comment (enola-dkfz) [reply src/db.rs:23]: done

File: src/db.rs (hunk 6 above)
Comment (enola-dkfz) [reply src/db.rs:32]: it was 8, now it's a CL arg, default 8

**Pull request merged**